polaris.shopify.com/pages/examples/index-table-with-nested-rows.tsx

```diff
@@ -0,0 +1,233 @@
+import {
+  LegacyCard,
+  Text,
+  useIndexResourceState,
+  IndexTable,
+} from '@shopify/polaris';
+import type {IndexTableRowProps, IndexTableProps} from '@shopify/polaris';
+import React, {Fragment} from 'react';
+import {withPolarisExample} from '../../src/components/PolarisExampleWrapper';
+
+export function WithNestedRowsExample() {
+  interface Product {
+    id: string;
+    quantity: number;
+    price: string;
+    size: string;
+    color: string;
+    image?: string;
+    disabled?: boolean;
+  }
+
+  interface ProductRow extends Product {
+    position: number;
+  }
+
+  interface ProductGroup {
+    id: string;
+    position: number;
+    products: ProductRow[];
+  }
+
+  interface Groups {
+    [key: string]: ProductGroup;
+  }
+
+  const rows: Product[] = [
+    {
+      id: '3411',
+      quantity: 11,
+      price: '$2,400',
+      size: 'Small Small Small',
+      color: 'Orange Orange Orange',
+    },
+    {
+      id: '2562',
+      quantity: 30,
+      price: '$975',
+      size: 'Medium Medium Medium',
+      color: 'Orange Orange Orange',
+    },
+    {
+      id: '4102',
+      quantity: 27,
+      price: '$2885',
+      size: 'Large',
+      color: 'Orange',
+    },
+    {
+      id: '2564',
+      quantity: 19,
+      price: '$1,209',
+      size: 'Small',
+      color: 'Red',
+      disabled: true,
+    },
+    {
+      id: '2563',
+      quantity: 22,
+      price: '$1,400',
+      size: 'Small',
+      color: 'Green',
+    },
+  ];
+
+  const columnHeadings = [
+    {title: 'Name', id: 'column-header--size'},
+    {
+      hidden: false,
+      id: 'column-header--price',
+      title: 'Price',
+    },
+    {
+      alignment: 'end',
+      id: 'column-header--quantity',
+      title: 'Available',
+    },
+  ];
+
+  const groupRowsByGroupKey = (
+    groupKey: keyof Product,
+    resolveId: (groupVal: string) => string,
+  ) => {
+    let position = -1;
+    const groups: Groups = rows.reduce((groups: Groups, product: Product) => {
+      const groupVal: string = product[groupKey] as string;
+      if (!groups[groupVal]) {
+        position += 1;
+
+        groups[groupVal] = {
+          position,
+          products: [],
+          id: resolveId(groupVal),
+        };
+      }
+      groups[groupVal].products.push({
+        ...product,
+        position: position + 1,
+      });
+
+      position += 1;
+      return groups;
+    }, {});
+
+    return groups;
+  };
+
+  const resourceName = {
+    singular: 'product',
+    plural: 'products',
+  };
+
+  const {selectedResources, allResourcesSelected, handleSelectionChange} =
+    useIndexResourceState(rows as unknown as {[key: string]: unknown}[], {
```

Comment:
I used this type conversion just for the sake of this example do not give typescript errors, as https://github.com/Shopify/polaris/blob/93defb11bd8e4c5f120265a0d3c0aad8a35f7771/polaris-react/src/utilities/use-index-resource-state.ts#L25 uses `useIndexResourceState<T extends {[key: string]: unknown}`